src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1748,5 +1747,54 @@ private void CheckContainsAnalyzerConfigDocuments(ImmutableArray<DocumentId> doc
                 CheckContainsAnalyzerConfigDocument(documentId);
             }
         }
+
+        /// <summary>
+        /// Throws if setting the project references of project <paramref name="projectId"/> to specified <paramref name="projectReferences"/>
+        /// would form a cycle in project dependency graph.
+        /// </summary>
+        private void CheckCircularProjectReferences(ProjectId projectId, IReadOnlyCollection<ProjectReference> projectReferences)
+        {
+            foreach (var projectReference in projectReferences)
+            {
+                if (projectId == projectReference.ProjectId || _state.ContainsTransitiveReference(projectReference.ProjectId, projectId))
+                {
+                    throw new InvalidOperationException(WorkspacesResources.The_project_already_transitively_references_the_target_project);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Throws if setting the project references of project <paramref name="projectId"/> to specified <paramref name="projectReferences"/>
+        /// would form an invalid submission project chain.
+        /// 
+        /// Submission projects can reference at most one other submission project. Regular projects can't reference any.
+        /// </summary>
+        private void CheckSubmissionProjectReferences(ProjectId projectId, IEnumerable<ProjectReference> projectReferences, bool ignoreExistingReferences)
```

Comment:
Then i'd need to get the project state in the caller as well. Seems simpler to have a bool, especially since this is a very specific helper used by two methods and not a general utility.